src/utils/SortingNetworkGenerator.java

```diff
@@ -1,129 +1,266 @@
+
+        boolean hasSameInput(Comparator other) {
+            return this.i1 == other.i1 ||
+                   this.i1 == other.i2 ||
+                   this.i2 == other.i1 ||
+                   this.i2 == other.i2;
         }
     }
 
-    public static boolean verifyPythonVersion() {
-        return (hasPython = hasPython // This caches the result if it's true
-            || verifyPythonVersion("3, 2, 0", "python3")
-            || verifyPythonVersion("3, 2, 0", "python")
-            || verifyPythonVersion("3, 2, 0", "py"));
+    private static final class WriterBuilderProxy {
+        final PrintWriter writer;
+        final StringBuilder builder;
+
+        WriterBuilderProxy(PrintWriter writer) {
+            this.writer = writer;
+            this.builder = null;
+        }
+
+        WriterBuilderProxy(StringBuilder builder) {
+            this.writer = null;
+            this.builder = builder;
+        }
+
+        void write(String s) {
+            if (builder == null)
+                writer.write(s);
+            else
+                builder.append(s);
+        }
+
+        @Override
+        public String toString() {
+            return builder == null ? writer.toString() : builder.toString();
+        }
+
+        String getValue() {
+            if (builder == null) {
+                throw new IllegalStateException("Cannot getValue() of PrintWriter");
+            }
+            return builder.toString();
+        }
     }
 
-    public static boolean verifyPythonVersionAndDialog() {
-        boolean hasVersion = verifyPythonVersion();
-        if (!hasVersion) {
-            JOptionPane.showMessageDialog(null, "It appears that you do not have Python 3.2 or later installed on your computer! Please install it before using this mode.",
-                "Sorting Network Visualizer", JOptionPane.WARNING_MESSAGE);
+    private static final int OUT_BUFFER_SIZE = 16_777_216; // 32 MB
+
+    private static int getMaxInput(Comparator[] comparators) {
+        int maxInput = 0;
+        for (Comparator c : comparators) {
+            if (c.i2 > maxInput) {
+                maxInput = c.i2;
+            }
         }
-        return hasVersion;
+        return maxInput;
     }
 
-    public static boolean encodeNetwork(int[] indices, String path) {
-        if (indices.length < 2) {
-            JOptionPane.showMessageDialog(null, "Sort does not compare indices; An empty sorting network cannot be generated.",
-                "File not saved", JOptionPane.ERROR_MESSAGE);
+    private static boolean encodeNetwork0(final Comparator[] comparators, final PrintWriter out) {
+        int scale = 1;
+        int xScale = scale * 35;
+        int yScale = scale * 20;
+        boolean small = comparators.length < 500_000;
 
-            return false;
+        int n = getMaxInput(comparators) + 1;
+        int h = (n + 1) * yScale;
+        double w = xScale;
+        Map<Comparator, Double> group = new HashMap<>();
```

Comment:
How would that work?

Reply:
seems java has no built-in Pair class, perhaps it could just be two parallel arrays of comparators and x-values? that'd probably be easier to work with, frankly

so then how it works now, just with keys being the first array and values being the second - I suggest this change since the map is never actually used for its primary purpose of getting values by keys.

Reply:
Oh! I get it now!

Reply:
Java does in fact have a builtin Pair class &mdash; `java.util.AbstractMap.SimpleImmutableEntry<K,V>`.